### PR TITLE
[RFC] Add portfolio choice to IndShk.

### PR DESCRIPTION
**Pull request by pkofod**
cc @llorracc @mnwhite 

This is a start of adding portfolio choice to the workhorse income shock/consumption/saving model. It basically does the "each period can itself be solved sequentially" trick.

Notes: 
 - RiskPremium  should be specified more freely, and specifically it should be controlled by an array!
 - Should we allow for correlation between risky assets and unemployment/transitory shocks?
 - <s>Shouldn't be called TradesStocks, it's risky asset more generally anyway</s>
 - <s> I've turned off two lines that do CubicInterpolation. This is because vFuncBool seems "too closely" tied to the CubicBool flag in the sense that as soon as vFuncBool is set to True, some parts of the code assumes that CubicBool is *also* set!</s>

I think it should be quite closely integrated in the existing infrastructure instead of a separate class that inherits. First, there's a solution for the perfect foresight type that we could "closed form" solve (so some things should go even earlier, all the way back to perf foresight model. Second, it makes for much less code repetition, as we can use the flag to control the methods that would otherwise be 95% identical across classes.

Edit:
Play around with it by c+p the following
```
from HARK.ConsumptionSaving.ConsIndShockModel import IndShockConsumerType
from copy import copy

CRRA = 2.0                          # Coefficient of relative risk aversion
Rfree = 1.01                        # Interest factor on assets
DiscFac = 0.96                      # Intertemporal discount factor
LivPrb = [0.98]                     # Survival probability
PermGroFac = [1.01]                 # Permanent income growth factor
AgentCount = 10000                  # Number of agents of this type (only matters for simulation)
aNrmInitMean = 0.0                  # Mean of log initial assets (only matters for simulation)
aNrmInitStd  = 1.0                  # Standard deviation of log initial assets (only for simulation)
pLvlInitMean = 0.0                  # Mean of log initial permanent income (only matters for simulation)
pLvlInitStd  = 0.0                  # Standard deviation of log initial permanent income (only matters for simulation)
PermGroFacAgg = 1.0                 # Aggregate permanent income growth factor (only matters for simulation)
T_age = None                        # Age after which simulated agents are automatically killed
T_cycle = 1                         # Number of periods in the cycle for this agent type
# Make a dictionary to specify a perfect foresight consumer type
init_perfect_foresight = { 'CRRA': CRRA,
                           'Rfree': Rfree,
                           'DiscFac': DiscFac,
                           'LivPrb': LivPrb,
                           'PermGroFac': PermGroFac,
                           'AgentCount': AgentCount,
                           'aNrmInitMean' : aNrmInitMean,
                           'aNrmInitStd' : aNrmInitStd,
                           'pLvlInitMean' : pLvlInitMean,
                           'pLvlInitStd' : pLvlInitStd,
                           'PermGroFacAgg' : PermGroFacAgg,
                           'T_age' : T_age,
                           'T_cycle' : T_cycle
                          }

# Parameters for constructing the "assets above minimum" grid
aXtraMin = 0.001                    # Minimum end-of-period "assets above minimum" value
aXtraMax = 20                       # Maximum end-of-period "assets above minimum" value
aXtraExtra = None                   # Some other value of "assets above minimum" to add to the grid, not used
aXtraNestFac = 3                    # Exponential nesting factor when constructing "assets above minimum" grid
aXtraCount = 48                     # Number of points in the grid of "assets above minimum"

# Parameters describing the income process
PermShkCount = 1                    # Number of points in discrete approximation to permanent income shocks
TranShkCount = 1                    # Number of points in discrete approximation to transitory income shocks
PermShkStd = [0.1]                  # Standard deviation of log permanent income shocks
TranShkStd = [0.1]                  # Standard deviation of log transitory income shocks
UnempPrb = 0.00                     # Probability of unemployment while working
UnempPrbRet = 0.000                 # Probability of "unemployment" while retired
IncUnemp = 1.0                     # Unemployment benefits replacement rate
IncUnempRet = 1.0                   # "Unemployment" benefits when retired
tax_rate = 0.0                      # Flat income tax rate
T_retire = 0                        # Period of retirement (0 --> no retirement)

# A few other parameters
BoroCnstArt = 0.0                  # Artificial borrowing constraint; imposed minimum level of end-of period assets
CubicBool = False                  # Use cubic spline interpolation when True, linear interpolation when False
vFuncBool = True                  # Whether to calculate the value function during solution

# Make a dictionary to specify an idiosyncratic income shocks consumer
init_idiosyncratic_shocks = { 'CRRA': CRRA,
                              'Rfree': Rfree,
                              'DiscFac': DiscFac,
                              'LivPrb': LivPrb,
                              'PermGroFac': PermGroFac,
                              'AgentCount': AgentCount,
                              'aXtraMin': aXtraMin,
                              'aXtraMax': aXtraMax,
                              'aXtraNestFac':aXtraNestFac,
                              'aXtraCount': aXtraCount,
                              'aXtraExtra': [aXtraExtra],
                              'PermShkStd': PermShkStd,
                              'PermShkCount': PermShkCount,
                              'TranShkStd': TranShkStd,
                              'TranShkCount': TranShkCount,
                              'UnempPrb': UnempPrb,
                              'UnempPrbRet': UnempPrbRet,
                              'IncUnemp': IncUnemp,
                              'IncUnempRet': IncUnempRet,
                              'BoroCnstArt': BoroCnstArt,
                              'tax_rate':0.0,
                              'vFuncBool':vFuncBool,
                              'CubicBool':CubicBool,
                              'T_retire':T_retire,
                              'aNrmInitMean' : aNrmInitMean,
                              'aNrmInitStd' : aNrmInitStd,
                              'pLvlInitMean' : pLvlInitMean,
                              'pLvlInitStd' : pLvlInitStd,
                              'PermGroFacAgg' : PermGroFacAgg,
                              'T_age' : T_age,
                              'T_cycle' : T_cycle,
}

init_lifecycle = copy(init_idiosyncratic_shocks)

init_lifecycle['CRRA'] = 6.0
init_lifecycle['PermGroFac'] = [1.01,1.01,1.01,1.01,1.01,1.02,1.02,1.02,1.02,1.02]
init_lifecycle['PermShkStd'] = [0,0,0,0,0,0,0,0,0,0]
init_lifecycle['TranShkStd'] = [0.3,0.2,0.1,0.3,0.2,0.1,0.3,0,0,0]
init_lifecycle['LivPrb']     = [0.99,0.9,0.8,0.7,0.6,0.5,0.4,0.3,0.2,0.1]
init_lifecycle['T_cycle']    = 10
init_lifecycle['T_retire']   = 7
init_lifecycle['T_age']      = 11 # Make sure that old people die at terminal age and don't turn into newborns!

### THIS IS THE PORTFOLIO PART ###
init_lifecycle['RiskPremium'] = 0.0
init_lifecycle['PortfolioBool'] = True

LifecycleExample = IndShockConsumerType(**init_lifecycle)

LifecycleExample.cycles = 1 # Make this consumer live a sequence of periods -- a lifetime -- exactly once
LifecycleExample.solve()

import numpy as np
import matplotlib.pyplot as plt
grid = np.linspace(0,10,140)
cons = LifecycleExample.solution[0].cFunc(grid)
plt.plot(grid, cons)
cons = LifecycleExample.solution[1].cFunc(grid)
plt.plot(grid, cons)
cons = LifecycleExample.solution[2].cFunc(grid)
plt.plot(grid, cons)
cons = LifecycleExample.solution[3].cFunc(grid)
plt.plot(grid, cons)
plt.show()
```

### CODE REVIEW

**Comment by llorracc** [comment]
@pkofod this looks good but please rewrite to use variable names consistent with the NARK guidelines and reissue the pull request when that revision is done

**Comment by pkofod** [comment]
> reissue the pull request when that revision is done

the reason why I closed the other PR was not that I made some changes and didn't reuse the other one, it was because it pointed to the wrong branch, and you cannot change that.

> please rewrite to use variable names consistent with the NARK guidelines

sure, I'll go through the variables and find the offenders.

**Comment by pkofod** [comment]
I've added some changes. I hope they were the ones you had identified yourself. I still think the bullet points in the original post is something we should discuss sooner rather than later.

**Comment by pkofod** [comment]
Of course `scripty.py` has to go before this is merged. It's just to have something easy to run :)

@mnwhite what's your take on injecting the functionality like this? This is very much related to many of the other discussions about inheritance vs one-class-to-rule-them-all. As mentioned above, it may even make sense to move this further up the line, to perf foresight model.

**Comment by pkofod** [comment]
@mnwhite would love some feedback here. I have pushed a totally re-worked version of this. I've tried to really be HARK-y, and used as much existing infrastructure as possible. There are some things I haven't fully implemented here, but that is because the Merton-Samuelson stuff is not yet in this branch. Those two need to be integrated, such that I can get the limiting solutions. In the script below, I have a quite high `aXtraMax` because of this. Below, you can see code I used to solve an instance.

```
import HARK.ConsumptionSaving.ConsPortfolioModel as cpm
import HARK.ConsumptionSaving.ConsumerParameters as param
import copy

RiskyDstnFunc = cpm.RiskyDstnFactory(RiskyAvg=1.08, RiskyStd=0.15)
init_portfolio = copy.copy(param.init_lifecycle)
init_portfolio['approxRiskyDstn'] = RiskyDstnFunc
init_portfolio['RiskyCount'] = 10
init_portfolio['RiskyShareCount'] = 100
init_portfolio['Rfree'] = 1.0
init_portfolio['CRRA'] = 6.0
init_portfolio['aXtraMax'] = 100
init_portfolio['aXtraCount'] = 100
init_portfolio['BoroCnstArt'] = 0.0 # important for theoretical reasons
pcct = cpm.PortfolioChoiceConsumerType(**init_portfolio)

pcct.solve()
import matplotlib.pyplot as plt
import numpy as np
eevalgrid = np.linspace(0,100,100)
plt.plot(eevalgrid, pcct.solution[6].RiskyShareFunc(eevalgrid))
plt.show()
```

It probably belongs in a more general discussion, but please look at what I did for the `makeLinearcFunc` method. One thing that makes it hard to extend some of the classes is that you need to overwrite quite a few methods that should really be "generic steps" in the solution algorithms. This method is a good example. It wants you to input `cNrm` and `mNrm` but not the limiting intercept and slope. This makes it a bit hard to reuse this code, as other solvers might also want to create such a cFunc, but the slope and intercept may no longer be known or has a different form. That's why I changed it to grab more "generic" fields with suffixes Intercept and Slope. That way I don't have to copy the whole method body into the PortfolioSolver class. The same goes for other steps in `prepare` methods, where a bit of refactoring could mean a whole lot less copy+pasting. As I said, this maybe belongs in a bigger discussion somewhere else.

**Comment by mnwhite** [comment]
I will take a look at this tomorrow (American) morning.  Can you send me a
reminder email around 9am eastern?

**Comment by pkofod** [comment]
Maybe I should drop the choice?

```
PortfolioChoiceConsumerType
LogNormalPortfolioChoiceConsumerType
``` 
vs
```
PortfolioConsumerType
LogNormalPortfolioConsumerType
```

**Comment by mnwhite** [comment]
Yeah, no need for Choice.

**Comment by pkofod** [comment]
I'm happy to take feedback from @mnwhite (and others) but it's in pretty good shape now I think. Things that could be added are mainly simulation. I also want to tweak the risky asset distribution interface, but that can also be done at a later time.

**Comment by llorracc** [comment]
@pkofod can you update this to fix the conflicts?

**Comment by llorracc** [comment]
One way of testing this would be to compare the model's solutions to those from the Campbell-Vicera approximation to portfolio choice in the Merton-Samuelson model, as is articulated (in Mathematica) in the lecture notes here:
[Portfolio-CRRA](http://www.econ2.jhu.edu/people/ccarroll/public/lecturenotes/AssetPricing/Portfolio-CRRA/)

An example of how this should work is in the relevant section of the [SolvingMicroDSOPs](http://www.econ2.jhu.edu/people/ccarroll/SolvingMicroDSOPs/#PlotRiskySharetOfat) lecture notes.

Patrick will make a DemARK with the tests he has already done (and will explain in his DemARK the things that originally confused him).

Mridul will work on getting these to the stage where they can be merged into HARK (for the basic tool) and the DemARK (for the tests/demonstrations).

PS.  Add a ".zip" to the url to get an archive with the LaTeX and the Mathematica code.

**Comment by MridulS** [comment]
@pkofod 
``` python
import HARK.ConsumptionSaving.ConsPortfolioModel as cpm
import HARK.ConsumptionSaving.ConsumerParameters as param
import copy

RiskyDstnFunc = cpm.RiskyDstnFactory(RiskyAvg=1.08, RiskyStd=0.15)
init_portfolio = copy.copy(param.init_lifecycle)
init_portfolio['approxRiskyDstn'] = RiskyDstnFunc
init_portfolio['RiskyCount'] = 10
init_portfolio['RiskyShareCount'] = 100
init_portfolio['Rfree'] = 1.0
init_portfolio['CRRA'] = 6.0
init_portfolio['aXtraMax'] = 100
init_portfolio['aXtraCount'] = 100
init_portfolio['BoroCnstArt'] = 0.0 # important for theoretical reasons
pcct = cpm.PortfolioConsumerType(**init_portfolio)

pcct.solve()
import matplotlib.pyplot as plt
import numpy as np
eevalgrid = np.linspace(0,100,100)
plt.plot(eevalgrid, pcct.solution[6].RiskyShareFunc(eevalgrid))
plt.show()
```
doesn't work as `RiskyShareFunc` is not a function, it's defined as a list of a list of `LinearInterp` (line 906 and 910), is this by design? To make this work the script should have `pcct.solution[6].RiskyShareFunc[0][0](eevalgrid)` instead.

**Comment by pkofod** [comment]
> doesn't work as `RiskyShareFunc` is not a function, it's defined as a list of a list of `LinearInterp` (line 906 and 910), is this by design? To make this work the script should have `pcct.solution[6].RiskyShareFunc[0][0](eevalgrid)` instead.

It is indeed by design...

`pcct.solutiont[t].RiskyShareFunc[can_adjust_state][past_state](eevalgrid)`

If the agent can adjust the second index doesn't matter, so that is `[0][0]` but if the agent can't adjust then you're in `[1][past_state]` territory, which basically means that if you cannot adjust, your investment function is basically the constant function with what-ever share you had going in.

I have to pull up my final script. Let me find that for you.

**Comment by pkofod** [comment]
@MridulS I believe this should be the final script. It solves, plots some optimal policies and some simulated data https://gist.github.com/pkofod/e24732ecd7a1be01e9def119bf5ea51f

File: HARK/ConsumptionSaving/ConsPortfolioModel.py [hunk 1]
```diff
@@ -0,0 +1,1062 @@
...
+
+   return -((1-rho)**-1)*np.dot(vals, weights)
+
+class PortfolioSolution(Solution):
+    distance_criteria = ['cFunc']
+
+    def __init__(self, cFunc=None, vFunc=None, wFunc=None,
+                       vPfunc=None, RiskyShareFunc=None, vPPfunc=None,
+                       mNrmMin=None, hNrm=None, MPCmin=None, MPCmax=None):
+        # Change any missing function inputs to NullFunc
```
[the remark this answers is not in the document]
Comment (pkofod) [reply HARK/ConsumptionSaving/ConsPortfolioModel.py]: > `MPCmin` and `MPCmax`

are not used currently, no. They could be, if/when someone verifies what they should be, and in that case they can be used to improve the consumption function's limiting behavior.

File: HARK/ConsumptionSaving/ConsPortfolioModel.py [hunk 2]
```diff
@@ -0,0 +1,1062 @@
...
+
+    def len(self):
+        return len(self.points)
+
+    def getPoints(self):
+        return self.points
+
+
+def calcwFunc(AdjustPrb, AdjustCount, ShareNowCount, vFunc, CRRA):
+    # AdjustCount cna be inferred from AdjustPrb...
```
[the remark this answers is not in the document]
Comment (pkofod) [reply HARK/ConsumptionSaving/ConsPortfolioModel.py]: > also we should change where functions are placed, they shouldn't be in between 2 classes and do we need these functions exposed to users of HARK? Like will users ever end up using `calcwFunc` directly in their code?

no



> The general way of dealing with these "helper" functions is to start them with an `_`, eg: `def _calcwFunc` and all of them are placed together in another file or top/bottom of the file (the underscore makes sure that they are not present in the final documentation, https://hark.readthedocs.io/en/latest/). IMO it makes the code much more readable.

that is fine

**Comment by pkofod** [comment]
@MridulS I made quite a lot of re-ordering, renaming and docs strings according to your recommendations. Could you have a look and also try the gist again (I've updated it slightly).

**Comment by llorracc** [comment]
Release Note: Added portfolio choice between risky and safe assets (ConsPortfolioModel)

**Pull request merged**